File: Onboard/OnboardingViewController.m
//
//  OnboardingViewController.m
//  Onboard
//
//  Created by Mike on 8/17/14.
//  Copyright (c) 2014 Mike Amaral. All rights reserved.
//

#import "OnboardingViewController.h"
#import "OnboardingContentViewController.h"
@import Accelerate;

static CGFloat const kPageControlHeight = 35;
static CGFloat const kBackgroundMaskAlpha = 0.6;
static CGFloat const kDefaultBlurRadius = 20;
static CGFloat const kDefaultSaturationDeltaFactor = 1.8;

@implementation OnboardingViewController

- (id)initWithBackgroundImage:(UIImage *)backgroundImage contents:(NSArray *)contents {
    self = [super init];

    // store the passed in background image and view controllers array
    _backgroundImage = backgroundImage;
    _viewControllers = contents;
    
    // we want the background masked by default
    _shouldMaskBackground = YES;
    
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // now that the view has loaded, we can generate the content
    [self generateView];
}

- (void)generateView {
    // create our page view controller
    _pageVC = [[UIPageViewController alloc] initWithTransitionStyle:UIPageViewControllerTransitionStyleScroll navigationOrientation:UIPageViewControllerNavigationOrientationHorizontal options:nil];
    _pageVC.view.frame = self.view.frame;
    _pageVC.view.backgroundColor = [UIColor whiteColor];
    _pageVC.delegate = self;
    _pageVC.dataSource = self;
    
    if (self.shouldBlurBackground) {
        [self blurBackground];
    }
    
    // create the background image view and set it to aspect fill so it isn't skewed
    UIImageView *backgroundImageView = [[UIImageView alloc] initWithFrame:self.view.bounds];
    backgroundImageView.contentMode = UIViewContentModeScaleAspectFill;
    [backgroundImageView setImage:_backgroundImage];
    [self.view addSubview:backgroundImageView];
    
    // as long as the shouldMaskBackground setting hasn't been set to NO, we want to
    // create a partially opaque view and add it on top of the image view, so that it
    // darkens it a bit for better contrast
    UIView *backgroundMaskView;
    if (self.shouldMaskBackground) {
        backgroundMaskView = [[UIView alloc] initWithFrame:_pageVC.view.frame];
        backgroundMaskView.backgroundColor = [UIColor colorWithWhite:0.0 alpha:kBackgroundMaskAlpha];
        [_pageVC.view addSubview:backgroundMaskView];
    }
    
    // more page controller setup
    [_pageVC setViewControllers:@[[_viewControllers firstObject]] direction:UIPageViewControllerNavigationDirectionForward animated:YES completion:nil];
    _pageVC.view.backgroundColor = [UIColor clearColor];
    [self addChildViewController:_pageVC];
    [self.view addSubview:_pageVC.view];
    [_pageVC didMoveToParentViewController:self];
    [_pageVC.view sendSubviewToBack:backgroundMaskView];
    [_pageVC.view sendSubviewToBack:backgroundImageView];
    
    // create and configure the the page control
    _pageControl = [[UIPageControl alloc] initWithFrame:CGRectMake(0, CGRectGetMaxY(self.view.frame) - kPageControlHeight, self.view.frame.size.width, kPageControlHeight)];
    _pageControl.numberOfPages = _viewControllers.count;
    [self.view addSubview:_pageControl];
}


#pragma mark - Convenience setters for content pages

- (void)setIconSize:(CGFloat)iconSize {
    for (OnboardingContentViewController *contentVC in _viewControllers) {
        contentVC.iconSize = iconSize;
    }
}

- (void)setFontName:(NSString *)fontName {
    for (OnboardingContentViewController *contentVC in _viewControllers) {
        contentVC.fontName = fontName;
    }
}

- (void)setTitleFontSize:(CGFloat)titleFontSize {
    for (OnboardingContentViewController *contentVC in _viewControllers) {
        contentVC.titleFontSize = titleFontSize;
    }
}

- (void)setBodyFontSize:(CGFloat)bodyFontSize {
    for (OnboardingContentViewController *contentVC in _viewControllers) {
        contentVC.bodyFontSize = bodyFontSize;
    }
}

- (void)setTopPadding:(CGFloat)topPadding {
    for (OnboardingContentViewController *contentVC in _viewControllers) {
        contentVC.topPadding = topPadding;
    }
}

- (void)setUnderIconPadding:(CGFloat)underIconPadding {
    for (OnboardingContentViewController *contentVC in _viewControllers) {
        contentVC.underIconPadding = underIconPadding;
    }
}

- (void)setUnderTitlePadding:(CGFloat)underTitlePadding {
    for (OnboardingContentViewController *contentVC in _viewControllers) {
        contentVC.underTitlePadding = underTitlePadding;
    }
}

- (void)setBottomPadding:(CGFloat)bottomPadding {
    for (OnboardingContentViewController *contentVC in _viewControllers) {
        contentVC.bottomPadding = bottomPadding;
    }
}


#pragma mark - Page view controller data source

- (UIViewController *)pageViewController:(UIPageViewController *)pageViewController viewControllerBeforeViewController:(UIViewController *)viewController {
    // return the previous view controller in the array unless we're at the beginning
    if (viewController == [_viewControllers firstObject]) {
        return nil;
    }
    else {
        NSInteger priorPageIndex = [_viewControllers indexOfObject:viewController] - 1;
        return _viewControllers[priorPageIndex];
    }
}

- (UIViewController *)pageViewController:(UIPageViewController *)pageViewController viewControllerAfterViewController:(UIViewController *)viewController {
    // return the next view controller in the array unless we're at the end
    if (viewController == [_viewControllers lastObject]) {
        return nil;
    }
    else {
        NSInteger nextPageIndex = [_viewControllers indexOfObject:viewController] + 1;
        return _viewControllers[nextPageIndex];
    }
}


#pragma mark - Page view controller delegate

- (void)pageViewController:(UIPageViewController *)pageViewController didFinishAnimating:(BOOL)finished previousViewControllers:(NSArray *)previousViewControllers transitionCompleted:(BOOL)completed {
    // if we haven't completed animating yet, we don't want to do anything because it could be cancelled
    if (!completed) {
        return;
    }
    
    // get the view controller we are moving towards, then get the index, then set it as the current page
    // for the page control dots
    UIViewController *viewController = [pageViewController.viewControllers lastObject];
    NSInteger newIndex = [_viewControllers indexOfObject:viewController];
    [_pageControl setCurrentPage:newIndex];
}

- (void)moveToPageForViewController:(UIViewController *)viewController {
    [_pageVC setViewControllers:@[viewController] direction:UIPageViewControllerNavigationDirectionForward animated:YES completion:nil];
    [_pageControl setCurrentPage:[_viewControllers indexOfObject:viewController]];
}


#pragma mark - Image blurring

- (void)blurBackground {
    // Check pre-conditions.
    if (_backgroundImage.size.width < 1 || _backgroundImage.size.height < 1) {
        NSLog (@"*** error: invalid size: (%.2f x %.2f). Both dimensions must be >= 1: %@", _backgroundImage.size.width, _backgroundImage.size.height, _backgroundImage);
        return;
    }
    if (!_backgroundImage.CGImage) {
        NSLog (@"*** error: image must be backed by a CGImage: %@", _backgroundImage);
        return;
    }
    
    UIColor *tintColor = [UIColor colorWithWhite:0.7 alpha:0.3];
    CGFloat blurRadius = kDefaultBlurRadius;
    CGFloat saturationDeltaFactor = kDefaultSaturationDeltaFactor;
    CGRect imageRect = { CGPointZero, _backgroundImage.size };
    UIImage *effectImage = _backgroundImage;
    
    BOOL hasBlur = blurRadius > __FLT_EPSILON__;
    BOOL hasSaturationChange = fabs(saturationDeltaFactor - 1.) > __FLT_EPSILON__;
    if (hasBlur || hasSaturationChange) {
        UIGraphicsBeginImageContextWithOptions(_backgroundImage.size, NO, [[UIScreen mainScreen] scale]);
        CGContextRef effectInContext = UIGraphicsGetCurrentContext();
        CGContextScaleCTM(effectInContext, 1.0, -1.0);
        CGContextTranslateCTM(effectInContext, 0, -_backgroundImage.size.height);
        CGContextDrawImage(effectInContext, imageRect, _backgroundImage.CGImage);
        
        vImage_Buffer effectInBuffer;
        effectInBuffer.data     = CGBitmapContextGetData(effectInContext);
        effectInBuffer.width    = CGBitmapContextGetWidth(effectInContext);
        effectInBuffer.height   = CGBitmapContextGetHeight(effectInContext);
        effectInBuffer.rowBytes = CGBitmapContextGetBytesPerRow(effectInContext);
        
        UIGraphicsBeginImageContextWithOptions(_backgroundImage.size, NO, [[UIScreen mainScreen] scale]);
        CGContextRef effectOutContext = UIGraphicsGetCurrentContext();
        vImage_Buffer effectOutBuffer;
        effectOutBuffer.data     = CGBitmapContextGetData(effectOutContext);
        effectOutBuffer.width    = CGBitmapContextGetWidth(effectOutContext);
        effectOutBuffer.height   = CGBitmapContextGetHeight(effectOutContext);
        effectOutBuffer.rowBytes = CGBitmapContextGetBytesPerRow(effectOutContext);
        
        if (hasBlur) {
            // A description of how to compute the box kernel width from the Gaussian
            // radius (aka standard deviation) appears in the SVG spec:
            // http://www.w3.org/TR/SVG/filters.html#feGaussianBlurElement
            //
            // For larger values of 's' (s >= 2.0), an approximation can be used: Three
            // successive box-blurs build a piece-wise quadratic convolution kernel, which
            // approximates the Gaussian kernel to within roughly 3%.
            //
            // let d = floor(s * 3*sqrt(2*pi)/4 + 0.5)
            //
            // ... if d is odd, use three box-blurs of size 'd', centered on the output pixel.
            //
            CGFloat inputRadius = blurRadius * [[UIScreen mainScreen] scale];
            NSUInteger radius = floor(inputRadius * 3. * sqrt(2 * M_PI) / 4 + 0.5);
            if (radius % 2 != 1) {
                radius += 1; // force radius to be odd so that the three box-blur methodology works.
            }
            vImageBoxConvolve_ARGB8888(&effectInBuffer, &effectOutBuffer, NULL, 0, 0, radius, radius, 0, kvImageEdgeExtend);
            vImageBoxConvolve_ARGB8888(&effectOutBuffer, &effectInBuffer, NULL, 0, 0, radius, radius, 0, kvImageEdgeExtend);
            vImageBoxConvolve_ARGB8888(&effectInBuffer, &effectOutBuffer, NULL, 0, 0, radius, radius, 0, kvImageEdgeExtend);
        }
        BOOL effectImageBuffersAreSwapped = NO;
        if (hasSaturationChange) {
            CGFloat s = saturationDeltaFactor;
            CGFloat floatingPointSaturationMatrix[] = {
                0.0722 + 0.9278 * s,  0.0722 - 0.0722 * s,  0.0722 - 0.0722 * s,  0,
                0.7152 - 0.7152 * s,  0.7152 + 0.2848 * s,  0.7152 - 0.7152 * s,  0,
                0.2126 - 0.2126 * s,  0.2126 - 0.2126 * s,  0.2126 + 0.7873 * s,  0,
                0,                    0,                    0,  1,
            };
            const int32_t divisor = 256;
            NSUInteger matrixSize = sizeof(floatingPointSaturationMatrix)/sizeof(floatingPointSaturationMatrix[0]);
            int16_t saturationMatrix[matrixSize];
            for (NSUInteger i = 0; i < matrixSize; ++i) {
                saturationMatrix[i] = (int16_t)roundf(floatingPointSaturationMatrix[i] * divisor);
            }
            if (hasBlur) {
                vImageMatrixMultiply_ARGB8888(&effectOutBuffer, &effectInBuffer, saturationMatrix, divisor, NULL, NULL, kvImageNoFlags);
                effectImageBuffersAreSwapped = YES;
            }
            else {
                vImageMatrixMultiply_ARGB8888(&effectInBuffer, &effectOutBuffer, saturationMatrix, divisor, NULL, NULL, kvImageNoFlags);
            }
        }
        if (!effectImageBuffersAreSwapped)
            effectImage = UIGraphicsGetImageFromCurrentImageContext();
        UIGraphicsEndImageContext();
        
        if (effectImageBuffersAreSwapped)
            effectImage = UIGraphicsGetImageFromCurrentImageContext();
        UIGraphicsEndImageContext();
    }
    
    // Set up output context.
    UIGraphicsBeginImageContextWithOptions(_backgroundImage.size, NO, [[UIScreen mainScreen] scale]);
    CGContextRef outputContext = UIGraphicsGetCurrentContext();
    CGContextScaleCTM(outputContext, 1.0, -1.0);
    CGContextTranslateCTM(outputContext, 0, -_backgroundImage.size.height);
    
    // Draw base image.
    CGContextDrawImage(outputContext, imageRect, _backgroundImage.CGImage);
    
    // Draw effect image.
    if (hasBlur) {
        CGContextSaveGState(outputContext);
        CGContextDrawImage(outputContext, imageRect, effectImage.CGImage);
        CGContextRestoreGState(outputContext);
    }
    
    // Add in color tint.
    if (tintColor) {
        CGContextSaveGState(outputContext);
        CGContextSetFillColorWithColor(outputContext, tintColor.CGColor);
        CGContextFillRect(outputContext, imageRect);
        CGContextRestoreGState(outputContext);
    }
    
    // Output image is ready.
    UIImage *outputImage = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    _backgroundImage = outputImage;
}

@end

File: Swift/OnboardSwift/OnboardingViewController.swift
//
//  OnboardingViewController.swift
//  OnboardSwift
//
//  Created by Mike on 9/12/14.
//  Copyright (c) 2014 Mike Amaral. All rights reserved.
//

import UIKit

class OnboardingViewController: UIViewController, UIPageViewControllerDataSource, UIPageViewControllerDelegate {
    let pageViewController: UIPageViewController = UIPageViewController(transitionStyle: UIPageViewControllerTransitionStyle.Scroll, navigationOrientation: UIPageViewControllerNavigationOrientation.Horizontal, options: nil)
    let pageControl: UIPageControl = UIPageControl()
    let kPageControlHeight: CGFloat = 35
    let kBackgroundMaskAlpha: CGFloat = 0.6
    var backgroundImage: UIImage?
    var contents: [OnboardingContentViewController] = []
    var shouldMaskBackground: Bool = true
    
    init(backgroundImage: UIImage?, contents: [OnboardingContentViewController]) {
        self.backgroundImage = backgroundImage
        self.contents = contents
        super.init(nibName: nil, bundle: nil)
    }

    required init(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.edgesForExtendedLayout = .None
        
        generateView()
    }
    
    func generateView() {
        pageViewController.delegate = self
        pageViewController.dataSource = self;
        pageViewController.view.frame = self.view.frame;
        
        // create the background image view and set it to aspect fill so it isn't skewed
        var backgroundImageView: UIImageView = UIImageView(image: backgroundImage)
        backgroundImageView.frame = self.view.frame
        backgroundImageView.contentMode = .ScaleAspectFit
        self.view.addSubview(backgroundImageView)
        
        var backgroundMaskView = UIView()
        if shouldMaskBackground {
            backgroundMaskView = UIView(frame: self.view.frame)
            backgroundMaskView.backgroundColor = UIColor(white: 0.0, alpha: kBackgroundMaskAlpha)
            self.view.addSubview(backgroundMaskView)
        }
        
        
        pageViewController.setViewControllers([contents[0]], direction: UIPageViewControllerNavigationDirection.Forward, animated: false, completion: nil)
        pageViewController.view.backgroundColor = UIColor.clearColor()
        self.addChildViewController(pageViewController)
        self.view.addSubview(pageViewController.view)
        pageViewController.didMoveToParentViewController(self)
        pageViewController.view.sendSubviewToBack(backgroundMaskView)
        pageViewController.view.sendSubviewToBack(backgroundImageView)
        
        pageControl.frame = CGRectMake(0, CGRectGetMaxY(self.view.frame) - kPageControlHeight, self.view.bounds.size.width, kPageControlHeight)
        pageControl.numberOfPages = contents.count;
        pageControl.autoresizingMask = .FlexibleWidth | .FlexibleTopMargin
        self.view.addSubview(pageControl)
    }
    
    // convenience setters for content pages
    
    func setIconSize(size: CGFloat) {
        for contentViewController in contents {
            contentViewController.iconSize = size
        }
    }
    
    func setTitleTextColor(color: UIColor) {
        for contentViewController in contents {
            contentViewController.titleTextColor = color
        }
    }
    
    func setBodyTextColor(color: UIColor) {
        for contentViewController in contents {
            contentViewController.bodyTextColor = color
        }
    }
    
    func setButtonTextColor(color: UIColor) {
        for contentViewController in contents {
            contentViewController.buttonTextColor = color
        }
    }
    
    func setFontName(fontName: String) {
        for contentViewController in contents {
            contentViewController.fontName = fontName
        }
    }
    
    func setTitleFontSize(size: CGFloat) {
        for contentViewController in contents {
            contentViewController.titleFontSize = size
        }
    }
    
    func setBodyFontSize(size: CGFloat) {
        for contentViewController in contents {
            contentViewController.bodyFontSize = size
        }
    }
    
    func setTopPadding(padding: CGFloat) {
        for contentViewController in contents {
            contentViewController.topPadding = padding
        }
    }
    
    func setUnderIconPadding(padding: CGFloat) {
        for contentViewController in contents {
            contentViewController.underIconPadding = padding
        }
    }
    
    func setUnderTitlePadding(padding: CGFloat) {
        for contentViewController in contents {
            contentViewController.underTitlePadding = padding
        }
    }
    
    func setBottomPadding(padding: CGFloat) {
        for contentViewController in contents {
            contentViewController.bottomPadding = padding
        }
    }
    
    // convenience methods
    
    func indexOfViewController(viewController: UIViewController) -> Int {
        var indexOfVC: Int = 0
        for (index, element) in enumerate(contents) {
            if element == viewController {
                indexOfVC = index
                break
            }
        }
        return indexOfVC
    }
    
    // PRAGMA: page view controller data source
    
    func pageViewController(pageViewController: UIPageViewController, viewControllerAfterViewController viewController: UIViewController) -> UIViewController? {
        let indexOfCurrentVC = indexOfViewController(viewController)
        return indexOfCurrentVC < contents.count - 1 ? contents[indexOfCurrentVC + 1] : nil
    }
    
    func pageViewController(pageViewController: UIPageViewController, viewControllerBeforeViewController viewController: UIViewController) -> UIViewController? {
        let indexOfCurrentVC = indexOfViewController(viewController)
        return indexOfCurrentVC > 0 ? contents[indexOfCurrentVC - 1] : nil
    }
    
    
    // PRAGMA: page view controller delegate
    
    func pageViewController(pageViewController: UIPageViewController!, didFinishAnimating finished: Bool, previousViewControllers: [AnyObject]!, transitionCompleted completed: Bool) {
        if !completed {
            return
        }
        
        let newViewController = pageViewController.viewControllers[0] as UIViewController
        pageControl.currentPage = indexOfViewController(newViewController)
    }
}